activerecord/lib/active_record/nested_attributes.rb

```diff
@@ -491,6 +491,8 @@ def assign_nested_attributes_for_collection_association(association_name, attrib
           attribute_ids.empty? ? [] : association.scope.where(association.klass.primary_key => attribute_ids)
         end
 
+        existing_records = existing_records.index_by { |record| record.id.to_s }
```

Comment:
Why the casting to string? Wouldn't it be better to cast to whatever the attribute type is? (Most likely Integer, unless you use UUID)

Reply:
Well, I just replicated searching that was done previously with `Enumerable#detect`. I'm guessing it was like that because integer primary keys will come as strings when submitted through a form, and just converting everything to strings would work with any primary key type.

Reply:
Hum, right, but I'm questioning whether coercing in the other direction might be faster. Integers are faster to compare and lookup in a hash, but then it might be slower to cast from string to the proper type than the other way around.

Aside from that, your patch make sense to me. The old algorithm has quite obviously a less than desirable complexity, and you patch obviously bring it back to more or less O(n).

It's late here, but I'll try to come back to your patch next week.